### DISCUSSION TITLE
Data manipultation after set creation

### DISCUSSION THREAD

**Discussion by anthonysirico** [pyg-team/pytorch_geometric#5651] [Q&A]
Is there a way to manipulate the data after I have already created the InMemoryDataset? After it was created I tried running `dataset[0].y = 1` and nothing happened. Note, that I purposefully left out the `y` to try to manipulate it after creation. 

BLUF: Can I add anything to the dataset using a loop?

```
for i in dataset.indeces():
    if x <= 0.5:
        dataset[i].y = 1
    else:
        dataset[i].y = 0
```

**Comment by EdisonLeeeee** [comment]
You cannot directly manipulate the data via indexing on `dataset`, as `dataset[i]` creates a copy of the data object and the operation is not inplaced. Alternatively, you can use the argument `transform` inherited from `InMemoryDataset`  to manipulate the dataset:
```python
def my_transform(data):
    if x <= 0.5:
        data.y = 1
    else:
        data.y = 0
    return data

dataset = Your_dataset(..., transform=my_transform)

```

**Reply by anthonysirico** [reply]
Thanks, you wouldnt happen to have a suggestion on how to sort the data? For example, if I have a set with an attribute called performance, how can I sort the data via `dataset.data.performance`?

**Reply by anthonysirico** [reply]
@EdisonLeeeee Also, the technique you suggested is not working unless I did something wrong:
```
def add_graph_label(data):
    for i in known.indices():
        if i <= len(known)*0.25:
            data.y = torch.tensor(np.asarray([1]), dtype=torch.int64)
        else:
            data.y = torch.tensor(np.asarray([0]), dtype=torch.int64)
    return data
known_ = ImbCircuitDataset('test_imb', split=0.25, transform=add_graph_label)
```
For some reason, all the y values are 0.

And of course I want to sort the set based on `known.data.performance` then based on index value give it a y of 0 or 1.

**Reply by EdisonLeeeee** [reply]
> Thanks, you wouldnt happen to have a suggestion on how to sort the data? For example, if I have a set with an attribute called performance, how can I sort the data via `dataset.data.performance`?

Did you mean that?
```python
performance = torch.tensor([data.performance for data in dataset]) # Suppose data.performance is a scalar
idx = torch.argsort(performance)
dataset = dataset[idx]
```

> Also, the technique you suggested is not working unless I did something wrong:

In your `add_graph_label` function, `data.y` is always 0 as there is always a case `i> len(known)*0.25` in the loop. Could you clarify what is `known.indices()` in your dataset? Or you can give a minimal example to make it clear.

**Reply by anthonysirico** [reply]
`known` is the name of the dataset representing the training set. I used `known.indices()` for `i` since it is `range(0,10812)` The overall goal is to create the `known` dataset, sort it based on `known.data.performance`, then split it 25/75, where the top 25 is 1 and the rest is 0. Then I have the `unknown` set, which is the test set.

**Reply by anthonysirico** [reply]
I tried your code for the sorting and it didn't work. Performance is a single value float tensor for each graph. 
```
known[0].performance = tensor([22.7112])

performance = torch.tensor([data.performance for data in known])
idx = torch.argsort(performance)
known_ = known[idx]
known_.data.performance
```
This is what I ran and did not sort it.

**Reply by rusty1s** [reply]
Not that `dataset.data` is not modified, but it instead creates a view on the underlying data with shuffled indices.
```python
print(known[0])
known_ = known[idx]
print(known_[0])
```
will yield different outputs. If you do not want to create a view but a copy instead, try
```python
known_ = known[idx].copy()
```
